src/main/java/org/pac4j/j2e/filter/CallbackFilter.java

```diff
@@ -77,13 +79,17 @@ protected void internalFilter(final HttpServletRequest request, final HttpServle
         }
         logger.debug("credentials: {}", credentials);
 
-        final CommonProfile profile = (CommonProfile) client.getUserProfile(credentials, context);
+        final UserProfile profile = client.getUserProfile(credentials, context);
         logger.debug("profile: {}", profile);
+        saveUserProfile(context, profile);
+        redirectToOriginallyRequestedUrl(context, response);
+    }
+
+    protected void saveUserProfile(final WebContext context, final UserProfile profile) {
```

Comment:
Make `saveUserProfile` a specific method to allow override (like `redirectToOriginallyRequestedUrl`).